Source/DafnyLanguageServer/Handlers/DafnyCodeActionHandler.cs

```diff
@@ -0,0 +1,161 @@
+      return new CommandOrCodeActionContainer();
+    }
+    var quickFixers = GetDafnyCodeActionProviders();
+    var fixesWithId = GetFixesWithIds(quickFixers, document, request).ToArray();
+
+    var documentUri = document.Uri.ToString();
+    documentUriToDafnyCodeActiones.AddOrUpdate(documentUri, _ => fixesWithId, (_, _) => fixesWithId);
+    var codeActions = fixesWithId.Select(fixWithId => {
+      CommandOrCodeAction t = new CodeAction {
+        Title = fixWithId.DafnyCodeAction.Title,
+        Data = new JArray(documentUri, fixWithId.Id),
+        Diagnostics = fixWithId.DafnyCodeAction.Diagnostics
+      };
+      return t;
+    }
+    ).ToArray();
+    return new CommandOrCodeActionContainer(codeActions);
+  }
+
+  private DafnyCodeActionProvider[] GetDafnyCodeActionProviders() {
+    return new List<DafnyCodeActionProvider>() {
+      new VerificationDafnyCodeActionProvider()
+    }.Concat(
+      DafnyOptions.O.Plugins.SelectMany(plugin =>
+        plugin is ConfiguredPlugin { Configuration: PluginConfiguration configuration } ?
+            configuration.GetDafnyCodeActionProviders() : new DafnyCodeActionProvider[] { })).ToArray();
+  }
+
+  public override Task<CodeAction> Handle(CodeAction request, CancellationToken cancellationToken) {
+    var command = request.Data;
+    if (command == null || command.Count() < 2 || command[1] == null) {
+      return Task.FromResult(request);
+    }
+
+    string? documentUri = command[0]?.Value<string>();
+    if (documentUri == null || !documentUriToDafnyCodeActiones.TryGetValue(documentUri, out var quickFixes)) {
+      return Task.FromResult(request);
+    }
+
+    int? id = command[1]?.Value<int>();
+    if (id == null) {
+      return Task.FromResult(request);
+    }
+
+    DafnyCodeActionWithId? selectedDafnyCodeAction = quickFixes.Where(pluginDafnyCodeAction => pluginDafnyCodeAction.Id == id)
+      .FirstOrDefault((DafnyCodeActionWithId?)null);
+    if (selectedDafnyCodeAction == null) {
+      return Task.FromResult(request);
+    }
+
+    return Task.FromResult(new CodeAction {
+      Edit = new WorkspaceEdit() {
+        DocumentChanges = new Container<WorkspaceEditDocumentChange>(
+          new WorkspaceEditDocumentChange(new TextDocumentEdit() {
+            TextDocument = new OptionalVersionedTextDocumentIdentifier() {
+              Uri = documentUri
+            },
+            Edits = new TextEditContainer(GetTextEdits(selectedDafnyCodeAction.DafnyCodeAction.GetEdits()))
+          }))
+      }
+    });
+  }
+
+  private IEnumerable<TextEdit> GetTextEdits(DafnyCodeActionEdit[] quickFixEdits) {
+    var edits = new List<TextEdit>();
+    foreach (var (range, toReplace) in quickFixEdits) {
+      edits.Add(new TextEdit() {
+        NewText = toReplace,
+        Range = range
+      });
+    }
+    return edits.ToArray();
+  }
+}
+
+public class VerificationDafnyCodeActionProviderInput : IDafnyCodeActionInput {
```

Comment:
Verification is obsolete. It dates from the time when I had the quick fixer be defined in Dafny core instead of DafnyLanguageServer
I renamed it `DafnyCodeActionInput`. And what do you mean by your question? This class uses `DafnyDocument` to implement `IDafnyCodeActionInput` (which is the interface plugins implementers are going to work with)

Reply:
To me a name ending in `Input` seems wrong.

When I read a function signature `SomeFunction(SomeFunctionInput x)` then the name `SomeFunctionInput` doesn't give me more information than what the rest of the function signature already did. What I'd like to know is what information can I get from `x`, and the name of the type of `x` should help with that.

Instead of the name `IDafnyCodeActionInput`, I think something like `IResolvedCompilation` or `IResolvedDocument` would tell me better what information I can get from it. If it's resolved then I expect that I can get the document text, the document AST, and the resolved types.